src/wavm/WAVMWasmModule.cpp

```diff
@@ -1003,86 +1008,171 @@ U32 WAVMWasmModule::mmapFile(U32 fd, U32 length)
     }
 
-    logger->debug("mmap - Growing memory from {} to {} pages",
-                  currentPageCount,
-                  newPageCount);
+    logger->debug(
+      "MEM - Growing memory from {} to {} pages", oldPages, newPages);
+
+    // Get offset of bottom of new range
+    auto newMemBase = (U32)(newMemPageBase * WASM_BYTES_PER_PAGE);
+
+    // Set current break to top of the new memory
+    currentBrk = getMemorySizeBytes();
 
-    // Get pointer to mapped range
-    auto mappedRangePtr = (U32)(Uptr(pageCountOut) * WASM_BYTES_PER_PAGE);
+    if (newMemBase != oldBytes) {
+        logger->error("Expected base of new region ({}) to be end of memory "
+                      "before growth ({})",
+                      newMemBase,
+                      oldBytes);
 
-    return mappedRangePtr;
+        throw std::runtime_error("Memory growth discrepancy");
+    }
+
+    if (currentBrk != newBytes) {
+        logger->error(
+          "Expected new brk ({}) to be old memory plus new bytes ({})",
+          currentBrk,
+          newBytes);
+        throw std::runtime_error("Memory growth discrepancy");
+    }
+
+    return newMemBase;
+}
+
+void WAVMWasmModule::shrinkMemory(U32 nBytes)
+{
+    auto logger = faabric::util::getLogger();
+
+    if (nBytes > currentBrk) {
+        logger->error(
+          "Shrinking by more than current brk ({} > {})", nBytes, currentBrk);
+        throw std::runtime_error("Shrinking by more than current brk");
+    }
+
+    U32 newBreak = currentBrk - nBytes;
+    if (!isWasmPageAligned(newBreak)) {
+        logger->error("New break not page aligned {}", newBreak);
+        throw std::runtime_error("New break not page aligned");
+    }
+
+    // Note - we don't actually free the memory, we just change the brk
+    logger->debug("MEM - shrinking memory {} -> {}", currentBrk, newBreak);
+    currentBrk = newBreak;
+}
+
+void WAVMWasmModule::unmapMemory(U32 offset, U32 nBytes)
+{
+    if (nBytes == 0) {
+        return;
+    }
+
+    auto logger = faabric::util::getLogger();
+
+    if (!isWasmPageAligned(offset) || !isWasmPageAligned(nBytes)) {
+        logger->error("Non-page aligned munmap {} {}", offset, nBytes);
+        throw std::runtime_error("Non-page aligned munmap");
+    }
+
+    faabric::util::UniqueLock lock(moduleMemoryMutex);
+
+    U64 maxPages = getMemoryType(defaultMemory).size.max;
+    U64 maxSize = maxPages * WASM_BYTES_PER_PAGE;
+
+    U32 unmapTop = offset + nBytes;
+    if (unmapTop > maxSize) {
+        logger->error(
+          "Munmapping outside memory max ({} > {})", unmapTop, maxSize);
+        throw std::runtime_error("munmapping outside memory max");
+    }
+
+    if (unmapTop == currentBrk) {
+        logger->debug("MEM - munmapping top of memory by {}", nBytes);
+        shrinkMemory(nBytes);
+    } else {
+        logger->warn("MEM - unable to reclaim memory {} at {}", nBytes, offset);
+    }
+}
+
+U32 WAVMWasmModule::mmapMemory(U32 nBytes)
+{
+    return growMemory(nBytes);
```

Comment:
Although we still provide an `mmap` call, it's translated into growing/ shrinking linear memory.